sympy/core/numbers.py

```diff
@@ -1286,26 +1286,25 @@ def __int__(self):
     __long__ = __int__
 
     def __eq__(self, other):
-        if isinstance(other, float):
-            # coerce to Float at same precision
-            o = Float(other)
-            try:
-                ompf = o._as_mpf_val(self._prec)
-            except ValueError:
-                return False
-            return bool(mlib.mpf_eq(self._mpf_, ompf))
         try:
             other = _sympify(other)
         except SympifyError:
             return NotImplemented
+        if not self:
+            return not other
         if other.is_NumberSymbol:
             if other.is_irrational:
                 return False
             return other.__eq__(self)
         if other.is_Float:
-            # precision is ignored so Float(.1) == Float(.1, 100)
-            # and Float(.1, 100) == Float(.1)
-            return bool(mlib.mpf_eq(self._mpf_, other._mpf_))
+            # compare at the same precision
```

Comment:
I would prefer the way this was above, comparison of unmodified mpf's. In particular, floats should not be truncated if we want them to compare as in Python. (And they should be sympified in `Float.__new__` by `mlib.from_float(num)`, with default precision and rounding. That would preserve mantissa and exponent.)

Reply:
I wrote it this way so `Float(1, 5) == Float(1) == Float(1, 50) == 1.0`. If we only want to preserve the last pair then all that would need to be done would be to compare always at `max(15, min(self._pre, other._prec))`...at least that's my first hunch.

Reply:
I was assuming that `mpf_eq` would do the "right thing" and return True for mpf's that represent the same rational number even if the mpf's were not exactly the same. I'll have to look into that more closely.

Reply:
I see now, `mpf_eq` is too simple-minded. The mpf's have to be normalised first. But that should be done in a lossless way, with sufficiently high precision (`prec >=self._mpf_.bc` for `self` and similarly for `other`) to avoid truncation.

Reply:
Remember that we are not changing the original numbers. 

I see that I should have said that I was trying to preserve `Float(0.1, 5) == Float(0.1) == Float(0.1, 50) == 0.1` If we don't want that then we will only have `Float(0.1) == 0.1`.

For reference,
```
>>> a, b = Decimal('0.100'), Decimal('0.1')
>>> a==b
True
>>> hash(a)==hash(b)
True
>>> set([a, b]) == set([a])
True
```

Reply:
BTW, all Floats are normalized in `Float._new` so if we want to keep pure comparisons (i.e. considering precision) then one can just compare the `._mpf_` directly.

Reply:
> Float(0.1, 5) == Float(0.1) == Float(0.1, 50) == 0.1

I don't think that this would be necessary. The denominator of 0.1 (as a rational number) is not a power of 2. Therefore it can only be approximated by Floats. Different precisions (5, 50) give different approximations and they are not expected to be equal. The situation is not comparable to Decimal whose arguments represent rational numbers with denominator a power of 10. (There does not seem to exist an analogous "binary decimal" representation for numbers whose denominator is an exact power of 2.)